### PR TITLE
[feat] PIP-335: Add support Oxia as a metadata store

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Add support for Oxia (https://github.com/streamnative/oxia) as a possible metadata store backed.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [X] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Review by mattisonchao — APPROVED** [review]

**Comment by nodece** [comment]
This PR seems to require a PIP, could you make a PIP?

**Review by lhotari — APPROVED** [review]
LGTM

File: pulsar-broker/pom.xml [hunk 1]
```diff
@@ -164,6 +164,12 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>io.streamnative.oxia</groupId>
+      <artifactId>oxia-testcontainers</artifactId>
+      <scope>test</scope>
+    </dependency>
```
Comment (BewareMyPower) [inline pulsar-broker/pom.xml:171]: Should you remove this dependency? There is no new test added to `pulsar-broker`.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

**Comment by BewareMyPower** [comment]
And I agree with @nodece that we need a PIP for introducing a new plugin to the broker side.

**Comment by merlimat** [comment]
Ok, I can create a PIP, similar to Etcd support.

**Comment by merlimat** [comment]
@nodece @BewareMyPower PIP created #22009  

https://lists.apache.org/thread/bmlo879sm1dd9wkgqzhb7fzlppdzm3yl

**Review by hangc0276 — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/oxia/OxiaMetadataStore.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+    @Override
+    protected CompletableFuture<Stat> storePut(
+            String path, byte[] data, Optional<Long> optExpectedVersion, EnumSet<CreateOption> options) {
+        CompletableFuture<Void> parentsCreated = createParents(path);
+        return parentsCreated.thenCompose(
+                __ -> {
+                    var expectedVersion = optExpectedVersion;
+                    if (expectedVersion.isPresent()
+                            && expectedVersion.get() != -1L
```
Comment (eolivelli) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/oxia/OxiaMetadataStore.java:188]: nit: is it better to have a constant for the magic value -1 ?

**Review by eolivelli — APPROVED** [review]
Overall LGTM

Looking forward to see it in action

**Review by BewareMyPower — APPROVED** [review]

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/oxia/OxiaMetadataStore.java (hunk 2 above)
Comment (merlimat) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/oxia/OxiaMetadataStore.java:188]: This is used across all plugins. The metadata store interface takes -1 as no version, following the ZK api. We should change for all plugins.

**Review by nodece — APPROVED** [review]
LGTM

File: pom.xml [hunk 3]
```diff
@@ -1152,6 +1153,18 @@ flexible messaging model and an intuitive client API.</description>
...
+        <groupId>io.streamnative.oxia</groupId>
+        <artifactId>oxia-client</artifactId>
+        <version>${oxia.version}</version>
+        <classifier>shaded</classifier>
+      </dependency>
+      <dependency>
+        <groupId>io.streamnative.oxia</groupId>
+        <artifactId>oxia-testcontainers</artifactId>
+        <version>${oxia.version}</version>
+      </dependency>
```
[the remark this answers is not in the document]
Comment (merlimat) [reply pom.xml:1166]: This is the `dependencyManagement` section, just to fix the version, it's not adding the dependency.

**Pull request merged**